### PR TITLE
Allow annotation processors to be run against groovy stubs during a joint compile

### PR DESCRIPTION
**Pull request by Sarev0k**
I've posted a [description this problem](https://groups.google.com/d/msg/gradle-dev/CvkFHRve3Sk/EkOU4liL7X8J) to gradle-dev earlier this week.

### CODE REVIEW

**Comment by bigdaz** [comment]
Thanks @Sarev0k. We'll take a look sometime soon.

**Comment by Sarev0k** [comment]
Hey, @bigdaz what are the chances that this will make it into the Gradle 2.3 release?

**Comment by bigdaz** [comment]
There's no chance it will make it into 2.3, sorry. We're well into the release process: just doing final review before the first RC.

**Comment by Sarev0k** [comment]
@bigdaz When do you think you might have the time to take a closer look at this?

I'd like to avoid missing the window to get this into the next Gradle version (2.4 I'm assuming) if possible.

**Comment by bigdaz** [comment]
I don't have the Groovy compiler knowledge required to review this, so it's waiting on review by @adammurdoch or @alkemist. But we have a bit of a PR backlog at the moment, and a lot of high priority work, so I can't promise any turnaround.

**Comment by Sarev0k** [comment]
@bigdaz When is the cutoff for getting changes into Gradle 2.4?  Are you guys planning on going through your PR backlog before the 2.4 cutoff?

It's been over two months since I submitted this PR, and I'm concerned that it's going to miss the 2.4 boat.

**Comment by alkemist** [comment]
I'll have an update about this PR in the next few days.

**Comment by Sarev0k** [comment]
Thanks @alkemist!

**Comment by alkemist** [comment]
So there are some issues that we need to resolve…

Firstly, there's no test coverage for this feature. The minimum requirement is an integration test that proves that this change actually achieves what it is supposed to.

There's a bigger problem here though in that the implementation as is introduces non trivial overhead for Groovy joint compiles as soon as an annotation processor is put on the classpath. I think it's going to be more likely that this is unnecessary cost than necessary. For example, if the annotation processor makes any structural changes then it is pointless to use on Groovy code. It's only metadata type annotations (like the ones Dagger uses) that are useful on Groovy code (and, they rely on implementation details of the current implementation of joint compilation). We are in a bit of a grey area as annotation processing on Groovy code isn't really a thing, but happens to work in some cases.

The way we want to solve this is to let the user be explicit about the behaviour that they want. If we know that Groovy code is intended to be subject to annotation processing, then we can do what your patch does and put all of the stubs through the compiler (and therefore annotation processing). Otherwise, we can assume that the processing only applies to Java code and therefore don't have to impose the extra compilation cost on users.

I need another day to map out the details of exactly what this would look like, but it's basically some new options on `CompileOptions` and `GroovyCompileOptions`. 

If you are interested in getting this change through, you could start on the integration testing in the meantime.

**Comment by Sarev0k** [comment]
Thanks for the feedback @alkemist.

I know I'm missing integration testing for this feature.  I wasn't quite sure where these tests live in your codebase and how they are executed.  Can you point me to an example groovy integration test for groovy compilation that I can use as an example?  Also, should I write my own annotation processor for this test, or introduce an external dependency on Dagger?

I'd also like to say that as a developer, if I'm putting an annotation processor on the classpath, I would expect that annotation processor to be run on all java sources, including Groovy stubs files.  If I didn't want an annotation processor to be run against all java files, I wouldn't have added it to my classpath.  Annotation processors are used for code generation, and I see no reason why my groovy code should be treated any differently than my java code in this respect.

I also believe the developer is being explicit in the behavior they want, by adding the annotation processor to their classpath.

I understand that you are concerned about the performance that this would cost large groovy joint compilation projects, that by happenstance have an annotation processor on their classpath, but don't annotate any of their groovy classes.  I agree with you, this is a concern.  But before we go designing a sub-optimal developer experience, don't you think we should collect metrics to get an idea of what effect this has on large groovy joint compilation projects?  If this is a marginal difference in compile time, I'd rather trigger this behavior based on the presence of the annotation processor on the classpath, as this is going to be what a developer expects coming from the java world.

Can you recommend a project that I can gather these metrics from?  Does gradleware maintain a large groovy project for performance testing?

Thanks again for your detailed feedback.

**Comment by alkemist** [comment]
> I know I'm missing integration testing for this feature. I wasn't quite sure where these tests live in your codebase and how they are executed. 

You'll want to add this to https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/integTest/groovy/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec.groovy#L27-27 so we test it with all the Groovy versions we test for and the different compilation modes.

> Also, should I write my own annotation processor for this test, or introduce an external dependency on Dagger?

It would be strongly preferable to not depend on Dagger for the test.

> 'd also like to say that as a developer, if I'm putting an annotation processor on the classpath, I would expect that annotation processor to be run on all java sources, including Groovy stubs files.

I don't think this holds. Most people don't know there are Java stub files.

> Annotation processors are used for code generation, and I see no reason why my groovy code should be treated any differently than my java code in this respect.

Because if the processor effectively mutates the class (e.g. adds a new method), then this all breaks down. The stub class will have the new method (so the Java code compiling against it will work), but the method will not actually be there at runtime.

> I also believe the developer is being explicit in the behavior they want, by adding the annotation processor to their classpath.

Not quite explicit enough, is the point. We don't know if they really want the stubs processed, and there are cases when they wouldn't.

> Can you recommend a project that I can gather these metrics from? Does gradleware maintain a large groovy project for performance testing?

We don't have Groovy compilation benchmarks at the moment, only Java and Scala.

Compiling groovy-core with a patched version and timing is probably not a bad option. Or just script up something to generate a few thousand Groovy source files to compile and some Java.

**Comment by Sarev0k** [comment]
> Because if the processor effectively mutates the class (e.g. adds a new method), then this all breaks down. The stub class will have the new method (so the Java code compiling against it will work), but the method will not actually be there at runtime.

I don't think this assertion is correct.  Annotation processors generate code, they don't mutate already existing code.  They do not add new methods, they can only create new classes.

There are some fringe annotation processors such as Lombok that rely on private JVM implementation specific APIs to mutate classes, but they are by far in the minority and can break in any future version of the JDK.  No one would expect these to work with Groovy code.

If you look at the [official APIs](http://docs.oracle.com/javase/7/docs/api/javax/annotation/processing/package-summary.html) you will find publicly documentated APIs that are available that let you mutate existing classes.

Based on this, I think it's completely justifiable for a developer to expect a java annotation processor to apply to annotations in their Groovy code.

I'd also like to add that last December I emailed @glaforge and @melix from the Groovy team about supporting annotation processors in Groovy, and they explained to me that running annotation processors against the java stubs that are created in a Groovy joint compile was their story for supporting Dagger on Android devices.  I'm willing to forward you this thread if you'd like.

**Comment by alkemist** [comment]
> Annotation processors generate code, they don't mutate already existing code.

I didn't realise they couldn't add new methods. Lombok was the most used process AFAIK and I didn't realise it wasn't using supported API.

If there's a decent chance that the annotations will be useful on Groovy code, then it's probably reasonable to always pay the compile price.

That reduces the bar for this feature to be just the test coverage. We'll need to cover the joint compile scenario and the scenario where there is no Java code.

**Comment by Sarev0k** [comment]
Thanks for this discussion @alkemist.  I really appreciate it.

[Here's a blog post](http://notatube.blogspot.com/2010/11/project-lombok-trick-explained.html) giving an overview of Lombok's trick for your reference.

Are there any deadlines coming up that I should be aware of?  I'm planning on adding the test coverage this week.

**Comment by alkemist** [comment]
If you can get the test coverage in this week, then there's a very good chance it will make 2.4. Any later than that and the likelihood reduces. 

Sorry, I can't be more specific at this time. We haven't quite finalised plans, but it is looking like we are going to bring 2.4 forward in the usual schedule. We definitely won't go into code freeze this week, but might any time from mid next week on.

**Comment by Sarev0k** [comment]
Hey @alkemist.  I've just added the test coverage you requested.

**Comment by Sarev0k** [comment]
Just added some more negative test cases to make sure the stub generation and compilation logic was working as expected.

**Comment by alkemist** [comment]
Excellent thanks.

I'll be digging in today or tomorrow, so unless there are issues it will be merged soon.

File: subprojects/plugins/src/integTest/resources/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec/compileWithAnnotationProcessor/settings.gradle [hunk 1]
```diff
@@ -0,0 +1,7 @@
+include ':'
+include ':SimpleAnnotationProcessor'
+// Integration tests don't copy directories that aren't named after a test to build/tmp/test files, where these build scripts are actually run from.
```
Comment (alkemist) [inline subprojects/plugins/src/integTest/resources/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec/compileWithAnnotationProcessor/settings.gradle]: Would you mind inlining this stuff? By this, I mean having the tests write this stuff to disk programmatically. It's easier to compose that way and it's also easier to see what the test is actually testing at a glance.

**Comment by alkemist** [comment]
Looks good @Sarev0k 

If you could change the tests so that they write the build under test to disk, instead of using the test resources as per my comment, I can merge it in for 2.4.

File: subprojects/plugins/src/integTest/resources/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec/compileWithAnnotationProcessor/settings.gradle (hunk 1 above)
Comment (Sarev0k) [inline subprojects/plugins/src/integTest/resources/org/gradle/groovy/compile/BasicGroovyCompilerIntegrationSpec/compileWithAnnotationProcessor/settings.gradle]: Looks like I was mistaken when I wrote this comment.

When I looked at the source of [TestResources.java](https://github.com/gradle/gradle/blob/master/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/TestResources.java#L56) I noticed that a shared directory is also copied with each test.

So I was able to dramatically simplify this logic by moving SimpleAnnotationProcessor into the shared directory, and then that directory is copied into the temporary directory to actually run the integration test.

Considering that the settings.gradle now looks like:
        include ':'
        include ':SimpleAnnotationProcessor'

Do you think it's still worth constructing these inline, within the test?

**Comment by Sarev0k** [comment]
@alkemist I think I may have done something better than what you were originally suggesting.

Do you still want me to generate the settings.gradle within the test?

**Comment by Sarev0k** [comment]
What I could do to give you the readability you want is remove the `compile project(':SimpleAnnotationProcessor')` from the resources build.gradle and apply it from the test instead.

**Comment by alkemist** [comment]
It's fine how it is now.

I have a strong personal preference for self containing the test and not using test resources. Not everyone on the team feels this way. Before your recent change, there was enough complexity in the test resources project to push the decision past the objective line, now it's only subjective and there's precedent for the way it is.

Will merge today.

**Comment by alkemist** [comment]
High quality contribution @Sarev0k .

Thanks for seeing it through.

**Pull request merged**